### PR TITLE
Update release workflow to checkout the given sha

### PR DESCRIPTION
**Pull request by zanieb**
As far as I can tell, the release workflow would just run on the latest commit from `main` then crash and burn if that did not match the given SHA? It seems much clearer to actually checkout the given SHA so we can publish a release after making additional commits to `main`.

### CODE REVIEW

**Comment by konstin** [comment]
(history note: we had it this way in an earlier draft but then changed it to the current solution, iirc mostly for convenience but i unfortunately can't find the original reasoning anymore)

File: .github/workflows/release.yaml [hunk 1]
```diff
@@ -7,12 +7,15 @@ on:
         description: "The version to tag, without the leading 'v'. If omitted, will initiate a dry run (no uploads)."
         type: string
       sha:
-        description: "Optionally, the full sha of the commit to be released"
+        description: "The full sha of the commit to be released. If omitted, the latest commit on the default branch will be used."
```
Comment (konstin) [inline .github/workflows/release.yaml:10]: do you want to add a sanity check that this is actually a commit on main?

**Review by konstin — APPROVED** [review]

**Comment by charliermarsh** [comment]
I think it was written this way as a sanity check: so you provided the version (like v0.0.288), the branch, and the expected SHA, and it failed if the SHA didn't match that of the branch.

**Comment by zanieb** [comment]
I'm not sure if https://github.com/astral-sh/ruff/pull/7279/commits/410551bae1d7da43b8a58edd2c004a9c93714322 will work with a shallow clone

**Pull request merged**